File: src/main/java/com/heekng/celloct/config/oauth/service/CustomOAuth2UserService.java
package com.heekng.celloct.config.oauth.service;

import com.heekng.celloct.config.oauth.dto.OAuthAttributes;
import com.heekng.celloct.config.oauth.dto.SessionMember;
import com.heekng.celloct.entity.Member;
import com.heekng.celloct.repository.MemberRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.oauth2.client.userinfo.DefaultOAuth2UserService;
import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest;
import org.springframework.security.oauth2.client.userinfo.OAuth2UserService;
import org.springframework.security.oauth2.core.OAuth2AuthenticationException;
import org.springframework.security.oauth2.core.user.DefaultOAuth2User;
import org.springframework.security.oauth2.core.user.OAuth2User;
import org.springframework.stereotype.Service;

import javax.servlet.http.HttpSession;
import java.util.Collections;

@RequiredArgsConstructor
@Service
public class CustomOAuth2UserService implements OAuth2UserService<OAuth2UserRequest, OAuth2User> {

    private final MemberRepository memberRepository;
    private final HttpSession httpSession;

    @Override
    public OAuth2User loadUser(OAuth2UserRequest userRequest) throws OAuth2AuthenticationException {
        OAuth2UserService<OAuth2UserRequest, OAuth2User> delegate = new DefaultOAuth2UserService();
        OAuth2User oAuth2User = delegate.loadUser(userRequest);

        String registrationId = userRequest.getClientRegistration().getRegistrationId();
        String userNameAttributeName = userRequest.getClientRegistration().getProviderDetails().getUserInfoEndpoint().getUserNameAttributeName();

        OAuthAttributes attributes = OAuthAttributes.of(registrationId, userNameAttributeName, oAuth2User.getAttributes());

        Member member = saveOrUpdate(attributes);

        httpSession.setAttribute("member", new SessionMember(member));

        return new DefaultOAuth2User(
                Collections.singleton(new SimpleGrantedAuthority(member.getRole().getKey())), attributes.getAttributes(), attributes.getNameAttributeKey()
        );
    }

    /**
     * Member entity로 변환한다.
     * 이미 존재하는 사용자인 경우 프로필 사진만 변경
     * 존재하지 않는 사용자는 신규로 가입한다.
     * @param attributes
     * @return
     */
    private Member saveOrUpdate(OAuthAttributes attributes) {
        Member member = memberRepository.findByEmail(attributes.getEmail())
                .map(entity -> entity.updatePicture(attributes.getPicture()))
                .orElse(attributes.toEntity());
        return memberRepository.save(member);
    }
}

File: src/main/kotlin/com/heekng/celloct/config/oauth/service/CustomOAuth2UserService.kt
package com.heekng.celloct.config.oauth.service

import com.heekng.celloct.config.oauth.dto.OAuthAttributes
import com.heekng.celloct.config.oauth.dto.SessionMember
import com.heekng.celloct.entity.Member
import com.heekng.celloct.repository.MemberRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.oauth2.client.userinfo.DefaultOAuth2UserService
import org.springframework.security.oauth2.client.userinfo.OAuth2UserRequest
import org.springframework.security.oauth2.client.userinfo.OAuth2UserService
import org.springframework.security.oauth2.core.user.DefaultOAuth2User
import org.springframework.security.oauth2.core.user.OAuth2User
import org.springframework.stereotype.Service
import java.util.*
import javax.servlet.http.HttpSession

@Service
class CustomOAuth2UserService @Autowired constructor(
    private val memberRepository: MemberRepository,
    private val httpSession: HttpSession
) : OAuth2UserService<OAuth2UserRequest, OAuth2User> {

    override fun loadUser(userRequest: OAuth2UserRequest?): OAuth2User {
        val delegate: OAuth2UserService<OAuth2UserRequest, OAuth2User> = DefaultOAuth2UserService()
        val oAuth2User = delegate.loadUser(userRequest)

        val registrationId = userRequest!!.clientRegistration.registrationId
        val userNameAttributeName = userRequest.clientRegistration.providerDetails.userInfoEndpoint.userNameAttributeName

        val attributes = OAuthAttributes.of(registrationId, userNameAttributeName, oAuth2User.attributes)

        val member = saveOrUpdate(attributes)

        httpSession.setAttribute("member", SessionMember.fixture(member))

        return DefaultOAuth2User(
            Collections.singleton(SimpleGrantedAuthority(member.role!!.key)), attributes.attributes, attributes.nameAttributeKey
        )
    }

    private fun saveOrUpdate(attributes: OAuthAttributes): Member {
        val member = memberRepository.findByEmail(attributes.email)?.updatePicture(attributes.picture) ?: attributes.toEntity()
        return memberRepository.save(member)
    }

}